### PR TITLE
ENH: fast Hankel transform

### PR DESCRIPTION
**Pull request by ntessore**
Continuing the discussion from #7310 cc @prisae @larsoner @mreineck.

This proposes a pure Python implementation of the FFTLog algorithm by Talman (1978) as given by Hamilton (2000). This PR focuses on implementing the discrete Hankel transform; the logarithmic DCT and DST can be added in a second PR once this is done.

The hard numerical core is implemented, tested, and working. But I have struggled quite a bit over the weekend to find a somewhat intuitive interface for the fast Hankel transform, as it has so many moving parts. The least bad option, to me, seems to be:

- Have only one discrete Hankel transform function `fht` that computes a discrete version of of the Hankel transform. (This is the `fht` routine in Hamilton's FFTLog package.) The input array can optionally be biased internally to improve the approximation of a continuous Hankel transform.
- Do not expose the basic FFTLog routine `fhtq` that computes the biased discrete Hankel transform.
- Have a function `fhtoffset` that computes the optimal value of the logarithmic shift of the output array for given transform parameters (this is the `krgood` routine in Hamilton's FFTLog package). I find `fhtshift` to be the better name but it clashes with `fftshift`.
- Have `fht` accept `offset` as an input parameter. The `fht` function will not automatically compute an optimal shift internally, the user has to call `fhtoffset` if they wish to construct a low-ringing transform.

Choices I am unsure about:

- If multiple transforms with the same parameters are to be performed, the user might strongly want to precompute the `fhtcoeff` array and use it for all subsequent transforms to improve performance. In that case, there could be an optional `coeff=` parameter to pass precomputed coefficients to `fht` and `ifht`. However, in that case the values of all other parameters of the transform must match. This could be added in the future.
- These functions take real inputs only. Should they be called `rfht` and `irfht`?
- Should `fhtoffset` be listed next to `fht` and `ifht` in the documentation, since it is quite fundamental to the algorithm working as expected?

Please consider this still a request for comments. Once we have figured out the remaining points, I will undraft this PR.

### CODE REVIEW

**Comment by prisae** [comment]
I think I would prefer to not give up on the Hamilton-`fht`. What about renaming:
- `fht`-> `fht`
- `fhtq` -> `_fht` (private fct called by `fht` and `fftlog`) [this is what you currently call `fht`, right?)
- `fftl` -> `fftlog` (next PR)

But as it is it would be fine too for me. I use almost exclusively the fftlog, not the fht.

**Comment by prisae** [comment]
I am linking here for reference the email of Andrew Hamilton giving his permission to include the code:
https://mail.python.org/pipermail/scipy-dev/2016-October/021576.html

(that was four years ago, time flies...)

**Comment by ntessore** [comment]
I suppose it makes sense to have the original `fht` if the scope is approximating the continuous transforms, instead of computing the discrete Hankel transforms of a logarithmically spaced periodic sequence (which isn't something that comes up very naturally). I'll make that change.

File: scipy/fft/_fftlog.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+        A *= np.exp(-bias*((j - j_c)*dln + offset))
+
+    return A
+
+
+def ifht(A, dln, mu, offset=0.0, bias=0.0):
+    r'''Compute the inverse fast Hankel transform.
+
+    Computes the discrete inverse Hankel transform of a logarithmically spaced
+    periodic sequence using the FFTLog algorithm [1]_, [2]_.
```
Comment (larsoner) [inline scipy/fft/_fftlog.py:143]: CircleCI treats warnings as errors, in this case searching "warning" in the build log showed:
```
/home/circleci/repo/build/testenv/lib/python3.7/site-packages/scipy/fft/_fftlog.py:docstring of scipy.fft.ifht:4: WARNING: Unknown target name: "1".
/home/circleci/repo/build/testenv/lib/python3.7/site-packages/scipy/fft/_fftlog.py:docstring of scipy.fft.ifht:4: WARNING: Unknown target name: "2".
```
You need `References` in this docstring for these to point to.

Comment (ntessore) [reply scipy/fft/_fftlog.py:143]: Too much copy and past. Fixed now, will push all changes in one go if anything else comes up.

**Comment by ntessore** [comment]
Another point to note: The transform is defined in the FFTLog convention

    A_FFTLog(k) = ∫ a(r) J_mu(kr) k dr

instead of the [usual definition of the Hankel transform](https://en.wikipedia.org/wiki/Hankel_transform)

    A(k) = ∫ a(r) J_mu(kr) r dr

This could be changed in any number of ways, most elementarily

    A(k) = k⁻¹ ∫ [a(r) r] J_mu(kr) k dr          (1)
         = k⁻² ∫ a(r) [kr J_mu(kr)] k dr         (2)

where (1) would involve multiplying the input sequence by `r` and performing the unbiased transform (the more classical approach), while (2) would involve computing a biased transform (should not change the behaviour of the transform since the input sequence is unchanged).

Personally, I would keep the definition as it is, because the logarithmic transform is "clean" without extra factors of `k` and `r`. I'm only bringing it up because it would be hard to change once rolled out.

**Comment by tylerjereddy** [comment]
@prisae @mreineck @peterbell10 any thoughts on the FFT code here?

File: scipy/fft/tests/test_fftlog.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        offset = fhtoffset(dln, mu, initial=offset, bias=bias)
+
+    A = fht(a, dln, mu, offset=offset, bias=bias)
+    a_ = ifht(A, dln, mu, offset=offset, bias=bias)
+
+    assert_allclose(a, a_)
+
+
+def test_fht_special_cases():
+    a = np.random.randn(64)
```
Comment (tylerjereddy) [inline scipy/fft/tests/test_fftlog.py:100]: I wonder if it is advisable to pin down the random seed here and in other new tests that use `random` infrastructure?

I also noticed that recent numpy docs suggest a preferred alternative to `randn` (Robert et al. probably have more to say about than than I would though)
image

Comment (rkern) [reply scipy/fft/tests/test_fftlog.py:100]: Yes, set a seed at the top of the function with `np.random.seed(3491349965)` (just so we don't have to track down weird intermittent failures when the PRNG throws you some odd edge case), and use `np.random.standard_normal()` instead of `randn()`. Tests can continue to use `np.random.seed()` and the `np.random` convenience functions to generate test data.

If you want to write a little more modern code, use `rng = np.random.RandomState(3491349965)` and call `rng.standard_normal()`, etc. methods instead of the `np.random` convenience functions.

With the way that you are using the data here, you could probably safely use `rng = np.random.default_rng(3491349965)`, but I'm not going to press it.

Comment (ilayn) [reply scipy/fft/tests/test_fftlog.py:100]: If this is the new way to set seed I'm quite happy to start using it actually. Previously it was buried down under the discussion hence sorry for my noise.

Comment (ntessore) [reply scipy/fft/tests/test_fftlog.py]: I've implemented the first proposed change. You could also introduce a `rng` fixture that does all of this in a sensible automatic manner with zero overhead for the test writer.

**Comment by ntessore** [comment]
Slightly off topic, but since it’s been a while, I would appreciate if a scipy maintainer could comment on whether it is actually desirable to have these other transforms (here Hankel) in `scipy.fft`.

If yes, I have a similar package for the fast Legendre transform (`flt` on PyPI) which could easily be integrated.

If no, then I will happily close this PR and implement the functionality as a standalone package as well.

**Comment by tylerjereddy** [comment]
> Slightly off topic, but since it’s been a while, I would appreciate if a scipy maintainer could comment on whether it is actually desirable to have these other transforms (here Hankel) in scipy.fft.

@ntessore Sounds on topic to me--for new additions we often try to send a message to the mailing list to get opinions because there will be more eyes on messages there. Maybe you could write up a short email to the mailing list describing your work here and asking if folks would be interested in having it in SciPy?

**Comment by prisae** [comment]
Might be worth to revisit the original mailing list answers from 2016 and 2017, threads:
- https://mail.python.org/pipermail/scipy-dev/2016-October/021567.html
- https://mail.python.org/pipermail/scipy-dev/2017-April/021939.html

Active folks in the threads included Jacob Vanderplas, Ralf Gommers, Joshua Wilson, Daπid

**Comment by tylerjereddy** [comment]
Oh great, I wasn't aware there was already some discussion.

**Comment by ntessore** [comment]
Thanks for linking the prior discussion. It does sound like the FFTLog functions were welcomed into the module, and not like there is a desire to keep the zoo of related transforms out.

I'll merge master in to try and get some more of the tests to pass, but this PR should be good to go then.

**Comment by mreineck** [comment]
> @prisae @mreineck @peterbell10 any thoughts on the FFT code here?

Sorry for the delay. I think this is a good addition, and very compact for the functionality it provides.

File: scipy/fft/__init__.py [hunk 3]
```diff
@@ -44,6 +44,15 @@
...
    idstn - N-D Inverse discrete sine transform
 
+Fast Hankel Transforms
+======================
+
+.. autosummary::
+    :toctree: generated/
+
+    fht - Fast Hankel transform
+    ifht - Inverse of fht
```
Comment (tylerjereddy) [inline scipy/fft/__init__.py:54]: The CI failure looks related I suppose?

```
Checking tutorial files at doc/source/tutorial/*.rst:

=========
scipy.fft
=========

Non-deprecated objects in __all__: 40
Objects in refguide: 38

ERROR: objects in scipy.fft.__all__ but not in refguide::

    fht
    ifht

This issue can be fixed by adding these objects to
the function listing in __init__.py for this module
```

Comment (ntessore) [reply scipy/fft/__init__.py:54]: Isn't this the function listing in __init__.py?

Comment (larsoner) [inline scipy/fft/__init__.py:54]: Locally this makes `python runtests.py -n --refguide-check --submodule fft` pass for me:
```suggestion
   fht - Fast Hankel transform
   ifht - Inverse of fht
```

**Review by larsoner — COMMENTED** [review]
Can you also add a bit of explanatory text / examples to `doc/source/tutorial/fft.rst`? Otherwise LGTM

**Comment by tylerjereddy** [comment]
I committed Eric's suggestion directly to see if it fixes the refguide issue in CI.

Looks like just one comment from Eric above to address and we should be good to merge before branching?

**Comment by ntessore** [comment]
I have added an introduction that at least states the definition of the transform. I don't really have the time at the moment to add more, but happy if anyone else wants to do so.

NB: I wrote it down straight here in a GitHub editor so I definitely need to have a look at it before merging once CI comes through.

**Comment by larsoner** [comment]
@ntessore :

https://31503-1460385-gh.circle-artifacts.com/0/html-scipyorg/tutorial/fft.html#fast-hankel-transform

**Comment by ntessore** [comment]
> @ntessore :
> 
> https://31503-1460385-gh.circle-artifacts.com/0/html-scipyorg/tutorial/fft.html#fast-hankel-transform

Thanks, fixed the maths errors and bad function reference. My "tutorial" is not great so I fully understand if you consider it a no-go.

**Comment by larsoner** [comment]
> My "tutorial" is not great so I fully understand if you consider it a no-go.

It's a good starting point :)

For some reason CircleCI refuses to build when you push commits, so I just pushed an empty one. Will merge once CIs come back happy, thanks @ntessore !

**Comment by prisae** [comment]
You all rock, this has come a very long way and survived even a complete change in fft-backend. Huge thank you @ntessore, @larsoner, and everyone else involved!

**Comment by ilayn** [comment]
first time contributor commits require us to click a button "Approve & Run". It's Github's fight with coin miners on CI machines stuff.

**Comment by larsoner** [comment]
> first time contributor commits require us to click a button "Approve & Run". It's Github's fight with coin miners on CI machines stuff.

Indeed I did this to activate GitHub actions, and it did active those. I've never seen that button affect CircleCI, though -- and it didn't seem to work when I clicked it in this case. I looked on CircleCI's interface and didn't see anything similar, though I wouldn't be surprised if it were a similar issue.

**Comment by rgommers** [comment]
This PR introduces a dependency on `scipy.special` in `scipy.fft`, which we didn't have before. Which is getting a little circular. I noticed because it broke tests for my `meson` branch - previously `fft` was completely standalone and tests were passing.

We should try and adopt lazy loading for this kind of thing, as described in [PEP 562](https://www.python.org/dev/peps/pep-0562/) and https://scientific-python.org/specs/spec-0001/.

**Comment by ntessore** [comment]
@rgommers I did check (at the zero loop level) whether `scipy.fft` was imported anywhere in `scipy.special` to prevent circular imports, and I only came across a docstring example. Apologies if I didn't go deep enough anything.

**Comment by rgommers** [comment]
No worries @ntessore, you didn't really break anything, this is more a long-term maintenance issue. Over time, more and more modules start to depend on each other, which is bad for modularity, import times, etc. It's not always easy to avoid though.

**Comment by WarrenWeckesser** [comment]
This is a great enhancement, a very belated *thanks* to all!

We don't have a strict policy *requiring* them, but I think everyone agrees that examples in docstrings are a great help to users, and we've been trying to ensure that we have examples in all our functions (see https://github.com/scipy/scipy/issues/7168).  If anyone has a chance, could we get some illustrative examples added to the docstrings of `fht`, `ifht` and `fhtoffset`?

**Comment by davidgontier** [comment]
Many thx for the fht function.

Unfortunately, I am not able no make it work. My main problem is the lack of examples (I cannot find one).

Here is a suggestion of example.
I know that the Fourier transform of a Gaussian is a Gaussian. In radial coordinates, this means that the Hankel transform of a Gaussian is a Gaussian (careful that this is the definition of the Hankel transform with the r dr integration -not k dr-).  

I would really like to see a short code checking this fact numerically... I am not able to do it...

For another more complete example, it is known that the eigenfunctions of the Hankel functions are related to the Laguerre polynomials, see e.g. Formula 4.20.3 of [Special Function and their Applications, Lebedev]...

As a sidenote: the doc reads "The input array a is a periodic sequence of length n,...". What does this mean? Also, the j index runs from 0 to n-1, or from 1 to n?

**Comment by prisae** [comment]
@davidgontier - The first thing that comes to my mind: Did you try it on logarithmically-spaced values? It will not work otherwise (it is, in fact, the entire purpose of it).

The best reference otherwise is Hamilton's website, unfortunately the maths doesn't render properly any longer: https://jila.colorado.edu/~ajsh/FFTLog/

You can find some examples here: https://pyfftlog.readthedocs.io/en/latest/examples (the parameters have different names there, they are like in the original FFTLog, whereas in the implementation in SciPy more intuitive/descriptive names were used).

I agree that adding some examples would be good. For now it might be insightful to look at the tests, which are examples of some sort too: https://github.com/scipy/scipy/blob/main/scipy/fft/tests/test_fftlog.py

**Comment by davidgontier** [comment]
@prisae Thx a lot for all the references. This is what I was looking for.

The example in https://pyfftlog.readthedocs.io/en/latest/examples/fftlogtest.html#sphx-glr-examples-fftlogtest-py is well explained. I guess the same example translated in fft.fht would be a nice addition to the scipy documentation (or a pointer to the page).
